### PR TITLE
fix: fetch release notes from self-hosted gitlab independent from url

### LINKED ISSUE #6704: Support of "self hosted gitlab" changelog
**What Renovate type are you using?**

self hosted with GitLab Self Hosted CE

**Describe the bug**
Considering this line of code :
https://github.com/renovatebot/renovate/blob/4f48cd1a4e0022e14266b070d433520fb805b277/lib/workers/pr/changelog/index.ts#L28

I think we can't get our changelogs from our self hosted gitlab CE because our url looks like this : `https://git.foo.bar`

**Relevant debug logs**

```
DEBUG: Repository URL does not match any known github hosts - skipping changelog retrieval (repository=mygroup/myproject)
       "manager": "regex",
       "depName": "myDep",
       "sourceUrl": "https://git.foo.bar/mygroup/mydep"
```

it uses the default choice : GitHub

**To Reproduce**

**Additional context**

### PR DESCRIPTION
**Pull request by casdevs**
## Changes:

this is a follow-up to the already merged PR #8336
and allows to also fetch release notes from self-hosted gitlab instances that do not contain "gitlab" in their hostname
 

## Context:

Closes #6704

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] Unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Here is a deep link to the missing coverage: https://app.codecov.io/gh/renovatebot/renovate/compare/8356/tree/lib/workers/pr/changelog/release-notes.ts

It seems to be complaining about the param `platform: string = PLATFORM_TYPE_GITHUB` in two locations

**Comment by viceice** [comment]
> Here is a deep link to the missing coverage: https://app.codecov.io/gh/renovatebot/renovate/compare/8356/tree/lib/workers/pr/changelog/release-notes.ts
> 
> It seems to be complaining about the param `platform: string = PLATFORM_TYPE_GITHUB` in two locations

This happens because the function is never called without `platform` arg in tests.

**Comment by casdevs** [comment]
Thanks for the hint.
I'll take care of it, just forgot to set the PR to draft

**Comment by casdevs** [comment]
ok, should be good to go now

File: lib/workers/pr/changelog/index.ts [hunk 1]
```diff
@@ -28,9 +29,9 @@ export async function getChangeLogJSON(
     let res: ChangeLogResult | null = null;
 
     if (
-      args.sourceUrl?.includes('gitlab') ||
-      (args.platform === 'gitlab' &&
-        new URL(args.sourceUrl).hostname === new URL(args.endpoint).hostname)
+      (args.platform === PLATFORM_TYPE_GITLAB &&
+        new URL(args.sourceUrl).hostname === new URL(args.endpoint).hostname) ||
```
Comment (rarkins) [inline lib/workers/pr/changelog/index.ts:33]: `new URL()` throws if its param is not a valid URL. Should we make a util function in `lib/util/url.ts` that does a try/catch and then returns `null` or perhaps `{}`, or can we be certain that `sourceUrl` will always be defined and valid at this point in the code?

Comment (casdevs) [reply lib/workers/pr/changelog/index.ts]: @rarkins: I fixed this, please review again

**Review by rarkins — COMMENTED** [review]
Is there a way to elegantly reduce the `platform === GITLAB || url.includes('gitlab')` type of checks?

**Comment by casdevs** [comment]
not sure, both checks are needed to support self-hosted gitlab instances as well as staying backwards-compatible (I did not want to modify/combine any of the options passed through from parent functions)

we could nest the condition into a helper function, but it has little benefit as we still have to pass all args separately

**Review by rarkins — CHANGES_REQUESTED** [review]
I have concerns that the widespread `PLATFORM === 'gitlab'` logic repeated throughout might accidentally break users who are running GitLab but looking up changelogs from non-gitlab. We need to enforce DRY here to make the code easier to check for this potential mistake

**Comment by casdevs** [comment]
the passed-through `platform` args all boil down to https://github.com/renovatebot/renovate/pull/8356/files#diff-b3eae2c101bf4a41ca740faad786dd7acf2c59632847a2530c25e69715ec9b43R350 where I repeated the already existing logic to decide whether to fetch from Gitlab or Github, i.e. `input.project.github` is used to decide about the platform

also, all existing tests for fetching from public Gitlab instances pass, so I don't think there is a problem with that

but we maybe could change set the new `platform` arg to `project.github && !url.includes('gitlab') ? 'github' : 'gitlab'` instead of doing the OR condition later on in different places...?

**Comment by rarkins** [comment]
You could perhaps call it `hostType` when you are passing it, because technically `platform` remains as whatever the git server is (github, gitlab, etc) while the host type for changelog lookups can vary. Or even `changelogPlatform` if you want to be more explicit

**Comment by casdevs** [comment]
full ACK on maintainability, so I renamed the new arg to `changelogPlatform` to make it more clear.

I'll see whether it is possible to make it more DRY, too.

**Comment by casdevs** [comment]
ok, please have a look if it is DRY enough now

**Comment by casdevs** [comment]
@rarkins: any chance this gets reviewed/merged shortly?

File: lib/workers/pr/changelog/release-notes.ts [hunk 2]
```diff
@@ -314,11 +339,19 @@ export async function addReleaseNotes(
     releaseNotes = await packageCache.get(cacheNamespace, cacheKey);
     // istanbul ignore else: no cache tests
     if (!releaseNotes) {
+      const changelogPlatform =
+        input.project.gitlab ||
+        input.project.apiBaseUrl?.includes('gitlab') ||
+        input.project.baseUrl?.includes('gitlab')
+          ? PLATFORM_TYPE_GITLAB
+          : PLATFORM_TYPE_GITHUB;
+
```
Comment (rarkins) [inline lib/workers/pr/changelog/release-notes.ts:348]: Isn't this separate logic from what is done earlier in a different file?

```
      (args.platform === PLATFORM_TYPE_GITLAB &&
        compareHosts(args.sourceUrl, args.endpoint)) ||
      args.sourceUrl?.includes('gitlab')
```

**Comment by viceice** [comment]
needs deconflicting

**Comment by rarkins** [comment]
Closing due to conflicts

**Pull request closed without merging**